Ronald's root ranking revisited

ceebo opened the pull request:
This a pull request for discussing the root ranking code of @syzygy1 from #840 

Code in this PR is largely taken from there but with some tweaks to make it compile 11 months later.

The only real change by me is to do with how the TB ranking information is fed into the search. I think it is nicer because changes to the search are minimal and it is very obvious that the search remains the same if no TB information is present.

The idea behind the patch can be summed up as:

1. TB code produces a ranking of all root moves.
2. Always keep the root move list sorted by TBRank followed by Score followed by PreviousScore.
3. Use the search only to differentiate between root moves that have the same TBRank.

Like the original root ranking code it should always be able to convert endings perfectly and doesn't always insist on DTZ optimal play. Most of the time it will rank by expected game result and only rank by DTZ in certain tricky situations. It should be MultiPV compatible and I think it is nicer than the post-processing way of doing things in the Natural TB branch.

I've done very little testing with this code. It is intended to be a non-functional change unless the root position is in the TBs. Also I have removed any of the code that changes the output score based on TB information just to keep things as simple as possible. Personally I like the sound of the Texel approach where scores for known draws are mapped into a small range around zero, cursed wins go into a higher range above and definite wins are given a large bonus, but I digress...

MichaelB7 commented:
👍 - very nice , was hoping somebody would pick this up.  I tried , but failed.

syzygy1 commented:
Including the TBRank in the root-move comparison operator is a good idea and makes the multipv fix even simpler.

I suspect (based on experience from my early SF patches) that not showing the TB score (and showing 0 TB probes) in a TB root position will confuse users quite a lot.

Eliminating in-search WDL probing if the root position is in the TBs but DTZ is not available probably decreases the chances of converting TB wins without DTZ (but it will make play look nicer for that case).

ceebo commented:
Thanks for the comments. I agree that taking out the TB score completely would be confusing but I was just trying to get feedback on the overall approach first of all. Also it never occurred to me that Threads.tb_hits() could be zero so merely adding rootMoves.size() seemed insignificant.

Regarding WDL probing, yes you are right. I forgot that the WDL tables are only probed after a zeroing move so in my head I was thinking that the root search would just return "Known win in 1 move" for every winning move. Now that I think about it again that is obviously nonsense.

syzygy1 commented:
It's not nonsense, but the reason that the rule50_count condition was added.

With an early_mate-like approach (not always taking the cutoff) one could reconsider probing after non-zeroing moves. It would only make sense in nodes where the parent is winning (to check that the current node indeed is losing, i.e. that the move just made was one of the winning moves). For best effect it should probably be limited to positions where non-zeroing moves can lose the win without immediately losing material. So... lots of things to consider and not something ripe for inclusion in SF master at this stage. Probing with the rule50_count condition to deal with WDL-only at the root seems to be a reasonable compromise for now. (Just relying on root_probe_wdl is a possible compromise too.)

RootMoves.size() is to suggest to the user that each rootmove is probed. Early TB patches outputted 1, which raised some questions. Of course no solution here is ideal; UCI is limiting.

syzygy1 commented:
As to the overall approach, I like it ;-).

Some criticism is possible: the PV output for a winning move might not actually be winning. But other approaches will have the same problem, unless they continue to probe within the search (as discussed in my previous comment). Of course it is normal that PVs are not entirely reliable. Still, at some point we might want to address this properly.

ceebo commented:
A more philosophical criticism of this patch is that now search() is no longer in total control of which move comes out on top. For example I'm sure it is possible to find positions where MultiPV=2 and mate in a long amount of moves is preferred to a mate in a shorter amount.

This effect is mitigated by the fact that TB ranking only goes by strict DTZ when the position is difficult and so mates in these kind of positions would be very hard for the engine to find in the first place.

If we want to rule out this issue absolutely then it should be possible to go back to searching all moves but find a way for mate scores to jump to the top regardless of TBRank. I guess this would be slightly more invasive but I haven't thought at all about the implementation.

syzygy1 commented:
Upgrading/downgrading the TB rank for a move in case the search returns a positive/negative mate score would probably fix this (very rare) problem. (One could test the fix by temporarily ranking strictly according to DTZ, as if a repetition has occurred.) The nice thing of a mate score returned by the search is that it takes into account repetitions and the fifty move rule, so it is reliable.

ceebo commented:
Yes. I believe that this patch has no downsides compared to the current master. Moreover, if some kind of method for mates to rise to the top was implemented then this patch would become completely orthogonal to the discussion about enforcing strict DTZ play and getting rid of the notorious has_repeated function. We would have a MultiPV compatible TB implementation and the number of things to argue about would be reduced...

syzygy1 commented:
You still need has_repeated, but I see no problem with it. We also don't remove regular repetition detection from Stockfish just because someone might find it tricky.

syzygy1 commented:
> Some criticism is possible: the PV output for a winning move might not actually be winning.

This could be solved by probing moves in the PV to make sure they are winning or drawing (no need to probe in other branches of the tree, as least as far as this cosmetic problem is concerned).

mcostalba commented:
@ceebo using root moves scoring to convey TB knowledge is a nice solution and maybe, once I have proved my NTB implementation is correct, I will play a bit with it.

But currently it suffers for 2 shortcomings, one just a possible bug, the other is deeper and is design choice.

1. Current TB rank scoring does not takes in account search mate scores, it always assumes TB rank to be higher then search, but this is not true if search returns a mate.

2. You score by DTZ, but this has no theoretical foundation IMO. Why you prefer a move with DTZ 6 instead of one with DTZ 9? If DTZ 6 is the minimum DTZ then this makes sense because it is ensured that no draw by repetition are possible following that path, but in all other cases it seems to me just an arbitrary choice.

To me TB moves should be ranked ad follow:

1. Shortest path
2. Not-shortest path but preserving the WDL score
3. Remaining moves

To this TB knowledge you can mix search knowledge in this way (from highest to lowest)

1. Mate scores (eventually double checked against DTZ)
2. TB score
3. Search score

syzygy1 commented:
@mcostalba 
Your description of what the patch does is not correct.

This is the meaning of different ranks:
1000 - all winning moves in normal circumstances
901-999 - winning moves by DTZ if a repetition has occurred earlier
900 - winning move with dtz + cnt50 == 100 (the critical case where in rare cases off by one may play a role in case of suboptimal play by the opponent)
899-500 - cursed wins by DTZ. Playing the move with lowest DTZ/highest rank seems to give best chances.
0 - all "real" drawing moves
500-999 - losses by DTZ if they are cursed or somewhat close to cursed (2 * dtz + cnt50 > 100, dtz + cnt50 > 100 meaning cursed).
-1000 - all regular losing moves (it's up to the search in non-TB swindle mode to pick the one making life hardest for the opponent).

Problem 1 only occurs in very exceptional situations (and the trivial fix has been pointed out).
In normal circumstances, all wins are ranked equally, so mates will come out on top automatically.
Only if we are in the rare repetition case or in the rare case with a dtz+cnt50 == 100 move could a move ranked lower than 1000 turn out to be a mate. The trivial fix, if required, is to then assign that move rank 1000, as pointed out above. But this will be extremely extremely rare and could only occur with multipv > 1.

syzygy1 commented:
For ranks 899-500, the original patch displayed a draw score running from +0.01 to +0.49 as the cursed win got closer to a real win. So multipv lines for those moves would be sorted correctly on decreasing score.

syzygy1 commented:
It would make sense to give higher priority to the winning move with lowest DTZ if all winning moves have relatively high DTZ.

To achieve that, simply assign moves with relatively high DTZ (or perhaps high cnt50 + dtz) rank 1000 - (dtz + cnt50). Moves with lower DTZ (if any) still get rank 1000. As a result, if all winning moves have high DTZ the search will pick a winning move from those with lowest DTZ.

If this were done, it would become a bit more important in the multipv > 1 to make sure that moves originally ranked below 1000 are reassigned rank 1000 in case a mate is found (because mate scores found by the search override the TB win score otherwise displayed).

So this ranking approach allows for flexibility. We can "program" the desired behaviour both when winning and when losing (by playing with the "2 * dtz + cnt50 > 100" condition).

syzygy1 commented:
> To me TB moves should be ranked as follow:
> 1. Shortest path
> 2. Not-shortest path but preserving the WDL score (all at the same level)
> 3. Remaining moves

From a "natural" point of view, preferring the move with lowest DTZ only really makes sense in the high-DTZ case. Otherwise, you are going to end up with strange sacrifices and weird pawn moves (and trying to avoid them at the last minute may then easily lead to draw-by-rep problems).

By the way, I am not claiming that "root ranking" is the only good approach. But it is simple, works well and is reasonably flexible. Other approaches are possible, but may easily get tricky to get right.

Consider multipv.
Every iteration of multipv is supposed to add the next best line to the lines previously found. What if the search returns a losing TB move as next best line if there are still drawing moves, or a drawing move if there are still winning moves, etc? Of course one could ignore the problem and just search multipv as normal, with normal search scores, and then overwrite all search scores with TB scores and re-sort the lines once the search is stopped, but to me that seems to be more a hack (confusing to users) than a proper solution.

ceebo commented:
I've been away for a few days and now want to respond to the points of @mcostalba 

"1. Current TB rank scoring does not takes in account search mate scores, it always assumes TB rank to be higher then search, but this is not true if search returns a mate."

I pushed a commit that fixes this by abandoning comparison using TBRank when one of the moves has a mate score. The commit actually reduces the number of lines. The search goes back to searching all moves regardless of TBRank.

"2. You score by DTZ, but this has no theoretical foundation IMO. Why you prefer a move with DTZ 6 instead of one with DTZ 9? If DTZ 6 is the minimum DTZ then this makes sense because it is ensured that no draw by repetition are possible following that path, but in all other cases it seems to me just an arbitrary choice."

I want the first move of the engine to always guarantee winning a won position. For the rest of the moves in the MultiPV > 1 case I am only really interested in the engine returning correct WDL scores. Therefore you are correct that ranking all of the moves by DTZ is a little bit arbitrary. Note that full DTZ ranking is only supposed to happen in difficult positions and so this would be quite rare anyway.

"To me TB moves should be ranked as follow:

1. Shortest path
2. Not-shortest path but preserving the WDL score (all at the same level)
3. Remaining moves"

I would not be particularly opposed to that. It could even be simplified to:

1. Shortest path
2. Remaining moves according to WDL

But I guess this would make the code (slightly) more complicated because finding DTZ optimal moves is inherently a two-pass process: one pass to calculate the best DTZ, one pass to assign a higher rank to the optimal moves. Just ranking by DTZ directly avoids this complication.

ceebo commented:
I pushed another commit that might behave as I am intending. The idea is to only increase bestValue at the root if the value is a mate score or the move has the best possible TBRank. Since now the root move list won't be ordered strictly by TBRank I have to find the best possible TBRank in a separate loop. Even if this is correct it is getting a bit uglier than I had hoped for.

ceebo commented:
I have updated this branch with the method of adding an extra sort that I stumbled across in the other thread. Also I made a trivial rewrite of the has_repeated function. Previously it was harder to see that the variable 'e' is actually decreasing by 1 per iteration. Also I dropped the st->pliesFromNull part because I didn't think it was necessary.

mcostalba commented:
Megred #1285 instead.

ceebo commented:
@mcostalba This patch is unrelated to Early Mate (which I think is a good patch) so I'm not sure why you use the word "instead".

mcostalba commented:
Sorry, I have to look it better then, but at this point is post SF 9 material, hopefully we will be able to refresh of all the TB code in a single and comprensive attempt more than through unrelated small patches....we will see :-)

For the moment I reopen this one.

ceebo commented:
@mcostalba No problem to leave this until after SF9. This patch is about how we handle moves at the root when the position is in the TB. Early Mate was about how we score TB nodes inside the tree when the root position is outside the TB.

Although I still like the general idea of this PR I am tired of seeing it sitting around so I prefer to close this and reopen (or start a new PR) later on if there is desire.
